appsec/src/extension/backtrace.c

```diff
@@ -0,0 +1,280 @@
+    zval *file = zend_hash_str_find(frame, "file", sizeof("file") - 1);
+    zval id;
+    ZVAL_LONG(&id, index);
+
+#ifdef TESTING
+    // In order to be able to test full path encoded everywhere lets set
+    // only the file name without path
+    char *file_name = strrchr(Z_STRVAL_P(file), '/');
+    Z_TRY_DELREF_P(file);
+    ZVAL_STRINGL(file, file_name + 1, strlen(file_name) - 1);
+#endif
+
+    array_init(datadog_backtrace_frame);
+    HashTable *datadog_backtrace_frame_ht = Z_ARRVAL_P(datadog_backtrace_frame);
+    zend_hash_add(datadog_backtrace_frame_ht, _frame_line, line);
+    zend_hash_add(datadog_backtrace_frame_ht, _frame_function, function);
+    zend_hash_add(datadog_backtrace_frame_ht, _frame_file, file);
+    zend_hash_add(datadog_backtrace_frame_ht, _id_key, &id);
+
+    Z_TRY_ADDREF_P(function);
+    Z_TRY_ADDREF_P(file);
+
+    return true;
+}
+
+void php_backtrace_to_datadog_backtrace(
+    // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+    zval *php_backtrace, zval *datadog_backtrace)
+{
+    if (Z_TYPE_P(php_backtrace) != IS_ARRAY) {
+        return;
+    }
+
+    HashTable *php_backtrace_ht = Z_ARRVAL_P(php_backtrace);
+    unsigned int frames_on_stack = zend_array_count(php_backtrace_ht);
+
+    unsigned int top = frames_on_stack;
+    unsigned int bottom = 0;
+    if (get_global_DD_APPSEC_MAX_STACK_TRACE_DEPTH() < frames_on_stack) {
+        top = (unsigned int)round(
+            (double)get_global_DD_APPSEC_MAX_STACK_TRACE_DEPTH() *
+            STACK_DEFAULT_TOP_PERCENTAGE);
+        bottom = get_global_DD_APPSEC_MAX_STACK_TRACE_DEPTH() - top;
+    }
+
+    array_init(datadog_backtrace);
+
+    HashTable *datadog_backtrace_ht = Z_ARRVAL_P(datadog_backtrace);
+
+    zval *tmp;
+    zend_ulong index;
+    if (top > 0) {
+        ZEND_HASH_FOREACH_NUM_KEY_VAL(php_backtrace_ht, index, tmp)
+        {
+            zval new_frame;
+
+            if (!php_backtrace_frame_to_datadog_backtrace_frame(
+                    tmp, &new_frame, index)) {
+                continue;
+            }
+
+            zend_hash_next_index_insert_new(datadog_backtrace_ht, &new_frame);
+            if (--top == 0) {
+                break;
+            }
+        }
+        ZEND_HASH_FOREACH_END();
+    }
+
+    if (bottom > 0) {
+        unsigned int position = frames_on_stack - bottom;
+        DD_FOREACH_FROM(php_backtrace_ht, 0, position, index)
+        {
+            tmp = _z;
+            zval new_frame;
+
+            if (!php_backtrace_frame_to_datadog_backtrace_frame(
+                    tmp, &new_frame, index)) {
+                continue;
+            }
+
+            zend_hash_next_index_insert_new(datadog_backtrace_ht, &new_frame);
+        }
+        ZEND_HASH_FOREACH_END();
+    }
+}
+
+void generate_backtrace(zend_string *id, zval *dd_backtrace)
+{
+    array_init(dd_backtrace);
+
+    if (!get_global_DD_APPSEC_STACK_TRACE_ENABLED() || !id) {
```

Comment:
Not sure if this should be checked here.

Reply:
Done